### PR TITLE
feat: CXINT-1851: PDF Invoices feature library

### PR DESCRIPTION
**Pull request by SherwinVarghese**
# PDF Invoices

This feature allows users to view invoices for their orders and download PDF invoices using `Invoices` APIs.
When the user selects an invoice, the backend returns a byte array, which is converted to PDF and downloaded.

# Screenshots of features

## Order Details Page
pdf invoices

## Downloaded Invoice
image

## E2E Cypress Tests
Screenshot 2023-09-07 at 7 15 02 PM

# Unit Test coverage
```
TOTAL: 30 SUCCESS
TOTAL: 30 SUCCESS
TOTAL: 30 SUCCESS

=============================== Coverage summary ===============================
Statements   : 94.44% ( 102/108 )
Branches     : 89.28% ( 25/28 )
Functions    : 90.9% ( 50/55 )
Lines        : 94.23% ( 98/104 )
================================================================================
```

# Epic Release Checklist
## General

- [x] Epic has been tested on a production-like server using Spartacus installation script
- [x] Epic has been tested in all supported browsers and devices (Check list on DoD page)
- [x] Epic has sufficient end to end (cypress) test coverage
- [x] Schematics code deprecations have been provided and tested
- [x] Epic complies with [Security best practices](https://sap.github.io/spartacus-docs/security-best-practices/)
- [x] Epic complies with [Accessibility best practices](https://sap.github.io/spartacus-docs/a11y-best-practices/)
- [x] Epic CSS supports directionality (if required)
- [x] Epic code scans have been conducted [Compliance scans for branches](https://wiki.one.int.sap/wiki/display/spar/Compliance+scans+for+branches)
- [x] Epic renders successfully in SSR

## Audits/reviews

- [x] Architecture review of the epic
- [x] Security and threat modeling audit/review from the Security team
- [x] Accessibility audit/review from the UX team

## New Library

If your epic will be introduced or introduces a new library:

- [x] Library provides schematics to install the library
- [x] Library modules can be lazy loaded appropriately
- [x] Library has been tested for a customer using Spartacus installation script

## Sample data

If the new feature requires new or updated sample data for a specific Commerce backend version: [N/A]

- [x] Sample data has been added to the sampledata repo and merged to the next version to be released
- [x] New sample data has been loaded and tested on a dev server
- [x] New sample data has been loaded and tested in our CI server

## Documentation

- [x] Documentation input is provided for it to be published on the doc site.

### CODE REVIEW

File: feature-libs/checkout/base/root/config/default-checkout-config.ts
Comment (giancorderoortiz) [inline feature-libs/checkout/base/root/config/default-checkout-config.ts:1]: why TODO ?

Comment (SherwinVarghese) [reply feature-libs/checkout/base/root/config/default-checkout-config.ts:1]: Removed it and reverted, was used for testing.

File: feature-libs/order/components/order-details/order-detail-reorder/order-detail-reorder.component.html [hunk 1]
```diff
@@ -12,4 +12,6 @@
       </div>
     </div>
   </ng-container>
+  <cx-invoices-list> </cx-invoices-list>
+  <!-- TODO: REMOVE AND REPLACE IT WITH CMS -->
```
Comment (pawelfras) [inline feature-libs/order/components/order-details/order-detail-reorder/order-detail-reorder.component.html]: Are you going to make such a change in this PR? If not, let's create a ticket and add the ticket number to the comment for better visibility.

File: feature-libs/order/components/order-details/order-details.module.ts [hunk 2]
```diff
@@ -65,6 +69,7 @@ const moduleComponents = [
     AddToCartModule,
     KeyboardFocusModule,
     IconModule,
+    PDFInvoicesComponentsModule, //TODO: remove it and replace with CMS
```
Comment (pawelfras) [inline feature-libs/order/components/order-details/order-details.module.ts]: Are you going to make such a change in this PR? If not, let's create a ticket and add the ticket number to the comment for better visibility.

File: feature-libs/pdf-invoices/README.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# PDF zInvoices
```
Comment (pawelfras) [inline feature-libs/pdf-invoices/README.md]: Is the letter  `z` added on purpose?

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  /* For Enum use in HTML */
+  ICON_TYPE = ICON_TYPE;
+
+  private PAGE_SIZE = 5; //Default page size
+
+  sortOptions: SortModel[];
+  sort = 'byInvoiceIdAsc';
+
+  protected sortMapping: { [key: string]: string } = {
+    byCreatedAtAsc: 'invoiceDate:asc', //TODO: update the sort code after the API is changed
```
Comment (pawelfras) [inline feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts]: Are you going to make such a change in this PR? If not, let's create a ticket and add the ticket number to the comment for better visibility.

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    byCreatedAtDesc: 'invoiceDate:desc', //TODO: update the sort code after the API is changed
```
Comment (pawelfras) [inline feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts]: Are you going to make such a change in this PR? If not, let's create a ticket and add the ticket number to the comment for better visibility.

File: feature-libs/pdf-invoices/core/services/pdf-invoices.service.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+  userId: string;
+  orderId: string;
+
+  constructor(
+    private routingService: RoutingService,
+    private userIdService: UserIdService,
+    protected pdfInvoicesConnector: PDFInvoicesConnector
+  ) {
+    this.subscriptions.add(
```
Comment (pawelfras) [inline feature-libs/pdf-invoices/core/services/pdf-invoices.service.ts:30]: Is there a place where subscriptions are unsubscribed? We need to make sure they won't lead to memory leaks.

File: projects/storefrontapp/src/environments/environment.ts [hunk 7]
```diff
@@ -25,11 +25,12 @@ export const environment: Environment = {
   occApiPrefix: '/occ/v2/',
   cds: buildProcess.env.CX_CDS ?? false,
   b2b: buildProcess.env.CX_B2B ?? false,
-  cdc: buildProcess.env.CX_CDC ?? false,
+  cdc: buildProcess.env.CX_CDC ?? true,
```
Comment (pawelfras) [inline projects/storefrontapp/src/environments/environment.ts]: Is this change on purpose? If not, please don't forget to reverse it.

**Review by pawelfras — COMMENTED** [review]
Hi, I went through the code and this is what I've noticed:

- there are few TODOs and I'm assuming they have to be done before the release
- e2e tests are missing
- unit tests are missing

As things listed above have to be handled before merging to the `develop-6.5.x` (FYI base of this PR is `develop-6.4.x`) branch, my suggestion is to create an epic branch for the PDF invoices feature and merge all the changes related to the feature instead of updating this PR. Having PRs with smaller chunks of code to be reviewed may simplify and speed up the process.

File: feature-libs/pdf-invoices/README.md (hunk 3 above)
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/README.md]: it is a typo, will remove it

File: projects/storefrontapp/src/environments/environment.ts (hunk 7 above)
Comment (SherwinVarghese) [reply projects/storefrontapp/src/environments/environment.ts]: noted, thank you

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+@Component({
+  selector: 'cx-invoices-list',
+  templateUrl: './invoices-list.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class InvoicesListComponent implements OnInit, OnDestroy {
+  /* For Enum use in HTML */
+  ICON_TYPE = ICON_TYPE;
+
+  private PAGE_SIZE = 5; //Default page size
```
Comment (pawelfras) [inline feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts]: `should:` This parameter looks like customers may want to overwrite this value. Let's change it to `protected`.

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        this.globalMessageService.add(
+          { key: 'pdfInvoices.featureNotEnabled' },
+          GlobalMessageType.MSG_TYPE_ERROR
+        );
+      }
+      return EMPTY;
+    })
+  );
+
+  private subscription = new Subscription();
```
Comment (pawelfras) [inline feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts]: `should:` Let's avoid using the `private` access modifier as customers may want to extend this component and use all fields/methods in their own way. Let's change it to `protected`

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+        .pipe(skip(1))
+        .subscribe(() =>
+          this.updateQueryParams({ fields: InvoicesFields.FULL })
+        )
+    );
+
+    this.getSortOptions();
+  }
+
+  private updateQueryParams(partialParams: InvoiceQueryParams) {
```
Comment (pawelfras) [inline feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts]: `should:` Let's avoid using the `private` access modifier as customers may want to extend this component and use all fields/methods in their own way. Let's change it to `protected`

File: feature-libs/pdf-invoices/core/connectors/pdf-invoices.connector.spec.ts
Comment (pawelfras) [inline feature-libs/pdf-invoices/core/connectors/pdf-invoices.connector.spec.ts:1]: Please add `done` callback to all test cases where async functionality is invoked.
For Example:
```
it('should call adapter when getInvoicesForOrder is invoked', (done) => {
    let result;
    pdfInvoicesConnector
      .getInvoicesForOrder(mockUserId, mockOrderId, mockInvoiceQueryParams)
      .pipe(take(1))
      .subscribe((res: any) => {
          result = res;
          done()
      });
    expect(result).toBeUndefined();
    expect(adapter.getInvoicesForOrder).toHaveBeenCalledWith(
      mockUserId,
      mockOrderId,
      mockInvoiceQueryParams
    );
  });

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+    component.queryParams$
+      .pipe(take(1))
+      .subscribe((value: InvoiceQueryParams) => (queryParams = value));
+    expect(queryParams).toEqual({
+      currentPage: 0,
+      pageSize: 5,
+      fields: InvoicesFields.FULL,
+      sort: 'invoiceId:asc',
+    });
+  });
```
Comment (pawelfras) [inline feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.spec.ts]:
```suggestion
  it('should read document list', (done) => {
    let orderInvoiceList: OrderInvoiceList = {};
    let queryParams: InvoiceQueryParams = {};
    component.invoicesList$
      .pipe(take(1))
      .subscribe((value: OrderInvoiceList) => 
      orderInvoiceList = value;
      done();
      );
    expect(orderInvoiceList).toEqual(mockOrderInvoiceList);
    expect(component.pagination).toEqual({
      currentPage: 0,
      pageSize: 5,
      totalPages: 4,
      totalResults: 16,
      sort: 'invoiceId:asc',
    });
    component.queryParams$
      .pipe(take(1))
      .subscribe((value: InvoiceQueryParams) => {
      queryParams = value; 
      done();
      });
    expect(queryParams).toEqual({
      currentPage: 0,
      pageSize: 5,
      fields: InvoicesFields.FULL,
      sort: 'invoiceId:asc',
    });
  });
```
Make similar change in other places where async functionality is called.

File: feature-libs/pdf-invoices/core/services/pdf-invoices.service.spec.ts
Comment (pawelfras) [inline feature-libs/pdf-invoices/core/services/pdf-invoices.service.spec.ts:1]: As in previous comments, please use the `done` callback in async places.

File: feature-libs/pdf-invoices/occ/adapters/occ-pdf-invoices.adapter.spec.ts [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+            }`
+        );
+      });
+
+      expect(mockReq.cancelled).toBeFalsy();
+      mockReq.flush(mockInvoicesList);
+      expect(mockReq.request.responseType).toEqual('json');
+    });
+
+    it(`should result in error when Error is thrown`, () => {
```
Comment (pawelfras) [inline feature-libs/pdf-invoices/occ/adapters/occ-pdf-invoices.adapter.spec.ts]: `should:` please use `done` callback.

File: feature-libs/pdf-invoices/root/model/pdf-invoices.model.ts
Comment (pawelfras) [inline feature-libs/pdf-invoices/root/model/pdf-invoices.model.ts:1]: `query:` Is it true that all of these fields are optional? Let's avoid optional parameters (if possible) for better type safety and less complicated type guards.

File: feature-libs/pdf-invoices/schematics/add-pdf-invoices/index_spec.ts [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+          continue;
+        }
+        // TODO: after 4.0: use this test, as we'll have synced versions between lib's and root package.json
+        // const expectedVersion = (peerDependencies as Record<
+        //   string,
+        //   string
+        // >)[toAdd];
+        const expectedDependency = dependencies[toAdd];
+        expect(expectedDependency).toBeTruthy();
+        // expect(expectedDependency).toEqual(expectedVersion);
```
Comment (pawelfras) [inline feature-libs/pdf-invoices/schematics/add-pdf-invoices/index_spec.ts]: `query:` why is this assertion commented?

File: projects/storefrontapp-e2e-cypress/cypress/e2e/vendor/requested-delivery-date/requested-delivery-date.e2e.cy.ts
Comment (pawelfras) [inline projects/storefrontapp-e2e-cypress/cypress/e2e/vendor/requested-delivery-date/requested-delivery-date.e2e.cy.ts:1]: `should:` Changes in this file seem not to be related to the feature. Let's move it to a separate ticket/PR to have a better history of changes.

File: projects/storefrontapp-e2e-cypress/cypress/e2e/vendor/s4om/s4om-schedule-lines.e2e.cy.ts
Comment (pawelfras) [inline projects/storefrontapp-e2e-cypress/cypress/e2e/vendor/s4om/s4om-schedule-lines.e2e.cy.ts:1]: `should:` Changes in this file seem not to be related to the feature. Let's move it to a separate ticket/PR to have a better history of changes.

File: projects/storefrontapp-e2e-cypress/cypress/helpers/vendor/requested-delivery-date/requested-delivery-date.ts [hunk 14]
```diff
@@ -22,7 +22,7 @@ import {
 import { waitForPage } from '../../checkout-flow';
 import * as alerts from '../../global-message';
 
-export const poNumber = '777';
+export const poNumber = '111';
```
Comment (pawelfras) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/vendor/requested-delivery-date/requested-delivery-date.ts]: `query:` Is this change related to the task?

**Review by pawelfras — COMMENTED** [review]
In addition to the comments for specific files - please make sure the installation process works as expected and the feature has been tested with the fresh Angular application

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts (hunk 8 above)
Comment (vigneshkrish8691) [inline feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts]: Is the page size configurable by the customer, or does it have a default value of 5?

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+  Subscription,
+} from 'rxjs';
+import { catchError, skip, switchMap, take, tap } from 'rxjs/operators';
+
+@Component({
+  selector: 'cx-invoices-list',
+  templateUrl: './invoices-list.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class InvoicesListComponent implements OnInit, OnDestroy {
```
Comment (vigneshkrish8691) [inline feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts:44]: Is it feasible for customers to customize or override this component?Please note that if there are private methods within the component, they will not have the ability to override them. Is it fine?

File: feature-libs/pdf-invoices/root/feature-name.ts [hunk 16]
```diff
@@ -0,0 +1,8 @@
+/*
+ * SPDX-FileCopyrightText: 2023 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export const PDF_INVOICES_FEATURE = 'pdfInvoices';
+export const PDF_INVOICES_CORE_FEATURE = 'pdfInvoicesCore';
```
Comment (vigneshkrish8691) [inline feature-libs/pdf-invoices/root/feature-name.ts:8]: Why core added ? Are you trying switch between core and integration?

File: projects/storefrontapp-e2e-cypress/cypress/e2e/vendor/s4om/s4om-schedule-lines.e2e.cy.ts
Comment (SherwinVarghese) [reply projects/storefrontapp-e2e-cypress/cypress/e2e/vendor/s4om/s4om-schedule-lines.e2e.cy.ts:1]: We try to reuse the helper methods from this E2E.

File: projects/storefrontapp-e2e-cypress/cypress/helpers/vendor/requested-delivery-date/requested-delivery-date.ts (hunk 14 above)
Comment (SherwinVarghese) [reply projects/storefrontapp-e2e-cypress/cypress/helpers/vendor/requested-delivery-date/requested-delivery-date.ts]: Yes, will refactor the variable into the corresponding ts file.

File: feature-libs/pdf-invoices/core/services/pdf-invoices.service.ts (hunk 6 above)
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/core/services/pdf-invoices.service.ts:30]: will  create an ngondestroy method. thanks

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts (hunk 9 above)
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts]: noted.

File: feature-libs/pdf-invoices/root/feature-name.ts (hunk 16 above)
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/root/feature-name.ts:8]: This is done for lazy loading of the feature module.

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts (hunk 15 above)
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts:44]: Yes Vignesh, changed the access types of the component methods to protected for extensibility

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts (hunk 8 above)
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.ts]: the paze size will be changed to protected, so it can be accessed by the extending class

File: feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.spec.ts (hunk 11 above)
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/components/invoices-list/invoices-list.component.spec.ts]: noted, thanks Pawel

File: feature-libs/pdf-invoices/core/services/pdf-invoices.service.spec.ts
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/core/services/pdf-invoices.service.spec.ts:1]: noted, thanks Pawel

File: feature-libs/pdf-invoices/occ/adapters/occ-pdf-invoices.adapter.spec.ts (hunk 12 above)
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/occ/adapters/occ-pdf-invoices.adapter.spec.ts]: noted, thanks Pawel

File: feature-libs/pdf-invoices/root/model/pdf-invoices.model.ts
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/root/model/pdf-invoices.model.ts:1]: Most of the parameters can be optional, I will try to make the fields shown on the UI as mandatory.

File: feature-libs/pdf-invoices/schematics/add-pdf-invoices/index_spec.ts (hunk 13 above)
Comment (SherwinVarghese) [reply feature-libs/pdf-invoices/schematics/add-pdf-invoices/index_spec.ts]: removed this assertion as it is not present in other feature modules too.

File: projects/storefrontapp-e2e-cypress/cypress/e2e/vendor/s4om/s4om-schedule-lines.e2e.cy.ts
Comment (SherwinVarghese) [reply projects/storefrontapp-e2e-cypress/cypress/e2e/vendor/s4om/s4om-schedule-lines.e2e.cy.ts:1]: added the ticket numbers to E2Es as per the new guidelines Pawel

Comment (pawelfras) [reply projects/storefrontapp-e2e-cypress/cypress/e2e/vendor/s4om/s4om-schedule-lines.e2e.cy.ts:1]: Sure, it's a good approach and I don't mind having the ticket numbers in E2Es, nevertheless, this change is not related to PDF Invoices. I suggest doing this in a separate PR.

Comment (pawelfras) [reply projects/storefrontapp-e2e-cypress/cypress/e2e/vendor/s4om/s4om-schedule-lines.e2e.cy.ts:1]: As we discussed, this time let's keep these ticket labels.

**Review by pawelfras — APPROVED** [review]
